bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxCommunicator.java

```diff
@@ -0,0 +1,521 @@
+                            logger.trace("Non expected message received exp:{}, recv: {}", expectedMessageNumbers,
+                                    receivedMessageType);
+
+                            // Non expected reply received
+                            if (outgoingMessage != null) {
+                                transmitFirst(outgoingMessage); // put message in queue again
+                                skipTransmit = true; // Skip the transmit on the next cycle to receive the panel message
+                            }
+                        }
+                    }
+                }
+
+                // Inform the listeners
+                if (incomingMessage != null && incomingMessage.isChecksumCorrect()) {
+                    for (CaddxPanelListener listener : listenerQueue) {
+                        listener.caddxMessage(this, incomingMessage);
+                    }
+                } else {
+                    logger.warn(
+                            "CaddxCommunicator.run() Received packet checksum does not match. in: {} {}, calc {} {}",
+                            incomingMessage.getChecksum1In(), incomingMessage.getChecksum2In(),
+                            incomingMessage.getChecksum1Calc(), incomingMessage.getChecksum2Calc());
+                }
+            }
+        } catch (IOException e) {
+            logger.debug("CaddxCommunicator.run() IOException. Stopping sender thread. {}", getSerialPortName());
+            Thread.currentThread().interrupt();
+        }
+
+        logger.warn("CaddxCommunicator.run() Sender thread stopped. {}", getSerialPortName());
+    }
+
+    /**
+     * Event handler to receive the data from the serial port
+     *
+     * @param SerialPortEvent serialPortEvent The event that occurred on the serial port
+     */
+    @Override
+    public void serialEvent(@Nullable SerialPortEvent serialPortEvent) {
+        if (serialPortEvent == null) {
+            return;
+        }
+
+        if (serialPortEvent.getEventType() == SerialPortEvent.DATA_AVAILABLE) {
+            logger.trace("Data receiving from the serial port");
+            if (protocol == CaddxProtocol.Binary) {
+                receiveInBinaryProtocol(serialPortEvent);
+            } else {
+                receiveInAsciiProtocol(serialPortEvent);
+            }
+        }
+    }
+
+    private int readByte(InputStream stream) {
+        try {
+            return stream.read();
+        } catch (IOException e) {
+            return -1;
+        }
+    }
+
+    private void receiveInBinaryProtocol(SerialPortEvent serialPortEvent) {
+        int b = 0;
+
+        // Read the start byte
+        if (!inMessage) // skip until 0x7E
+        {
+            b = 0;
+            while (b != 0x7E && b != -1) {
+                b = readByte(in);
+            }
+            if (b == -1) {
+                return;
+            }
+
+            inMessage = true;
+            messageBufferLength = 0;
```

Comment:
Almost identical logic is used on line 449, can you refactor it?

Reply:
Done